src/client/streamableHttp.ts

```diff
@@ -0,0 +1,370 @@
+  private _activeStreams: Map<string, ReadableStreamDefaultReader<Uint8Array>> = new Map();
+  private _abortController?: AbortController;
+  private _url: URL;
+  private _requestInit?: RequestInit;
+  private _authProvider?: OAuthClientProvider;
+  private _sessionId?: string;
+  private _lastEventId?: string;
+
+  onclose?: () => void;
+  onerror?: (error: Error) => void;
+  onmessage?: (message: JSONRPCMessage) => void;
+
+  constructor(
+    url: URL,
+    opts?: StreamableHTTPClientTransportOptions,
+  ) {
+    this._url = url;
+    this._requestInit = opts?.requestInit;
+    this._authProvider = opts?.authProvider;
+  }
+
+  private async _authThenStart(): Promise<void> {
+    if (!this._authProvider) {
+      throw new UnauthorizedError("No auth provider");
+    }
+
+    let result: AuthResult;
+    try {
+      result = await auth(this._authProvider, { serverUrl: this._url });
+    } catch (error) {
+      this.onerror?.(error as Error);
+      throw error;
+    }
+
+    if (result !== "AUTHORIZED") {
+      throw new UnauthorizedError();
+    }
+
+    return await this._startOrAuth();
+  }
+
+  private async _commonHeaders(): Promise<HeadersInit> {
+    const headers: HeadersInit = {};
+    if (this._authProvider) {
+      const tokens = await this._authProvider.tokens();
+      if (tokens) {
+        headers["Authorization"] = `Bearer ${tokens.access_token}`;
+      }
+    }
+
+    if (this._sessionId) {
+      headers["mcp-session-id"] = this._sessionId;
+    }
+
+    return headers;
+  }
+
+  private async _startOrAuth(): Promise<void> {
+    try {
+      // Try to open an initial SSE stream with GET to listen for server messages
+      // This is optional according to the spec - server may not support it
+      const commonHeaders = await this._commonHeaders();
+      const headers = new Headers(commonHeaders);
+      headers.set('Accept', 'text/event-stream');
+
+      // Include Last-Event-ID header for resumable streams
+      if (this._lastEventId) {
+        headers.set('last-event-id', this._lastEventId);
+      }
+
+      const response = await fetch(this._url, {
+        method: 'GET',
+        headers,
+        signal: this._abortController?.signal,
+      });
+
+      if (response.status === 405 || response.status === 404) {
+        // Server doesn't support GET for SSE, which is allowed by the spec
+        // We'll rely on SSE responses to POST requests for communication
+        return;
+      }
+
+      if (!response.ok) {
+        if (response.status === 401 && this._authProvider) {
+          // Need to authenticate
+          return await this._authThenStart();
+        }
+
+        const error = new Error(`Failed to open SSE stream: ${response.status} ${response.statusText}`);
+        this.onerror?.(error);
+        throw error;
+      }
+
+      // Successful connection, handle the SSE stream as a standalone listener
+      const streamId = `initial-${Date.now()}`;
```

Comment:
Why do we need these stream IDs? I can't find any spot where we locate a stream by its ID.